### PR TITLE
rgw_file: add compression interop to RGW NFS

### PR DESCRIPTION
**Pull request by mattbenjamin**
The top commit was needed to test--I think there may be a regression here in building the compression plugins, but apologies if this is operator error.

### CODE REVIEW

**Comment by aclamk** [comment]
@mattbenjamin I had problem with libceph_zlib library. wip_nfs_compress@https://github.com/aclamk/ceph.git addresses it.

**Comment by mattbenjamin** [comment]
Hi Adam,

Ok, so this is an intrusive change and is going to need some discussion on
a PR.  If you agree, could you approve my original NFS compression PR and
then send this as a separate PR to start that discussion off?

thanks!

Matt

File: src/rgw/rgw_file.cc [hunk 1]
```diff
@@ -1286,6 +1286,27 @@ namespace rgw {
...
     op_ret = processor->prepare(get_store(), NULL);
+    if (op_ret < 0) {
+      ldout(s->cct, 20) << "processor->prepare() returned ret=" << op_ret
+			<< dendl;
+      goto done;
+    }
+
+    filter = processor;
+
+    if (compression_type == "none")
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.cc]: @aclamk seems harmless, and is concise

File: src/rgw/rgw_file.cc [hunk 2]
```diff
@@ -1344,6 +1365,12 @@ namespace rgw {
 	goto done;
       }
 
+      /* restore compression filter, if any */
+      if (compressor) {
+	compressor.emplace(s->cct, plugin, filter);
+	filter = &*compressor;
+      }
+
       op_ret = put_data_and_throttle(processor, data, ofs, false);
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.cc]: @aclamk good catch!

File: src/rgw/rgw_file.cc (hunk 1 above)
Comment (mattbenjamin) [inline src/rgw/rgw_file.cc]: @aclamk I guess since I do it every exec_start(), can just be a local

**Comment by aclamk** [comment]
@mattbenjamin I verified that your changes to zlib linking does not affect compression in radosgw. I assume that problems will only show when ceph_globals are defined in loaded so.

**Comment by mattbenjamin** [comment]
@aclamk , @cbodley I'm guessing I need to rebase this to remove at least the linkage to libglobal, since that appears to have been merged;  what about libcommon?  If that's not the case, can we merge this now?

**Comment by aclamk** [comment]
@mattbenjamin After #16245 there is no longer any need to modify linkage.

**Comment by mattbenjamin** [comment]
@aclamk thanks, updated

**Pull request merged**